mozilla-release/mobile/android/base/java/com/cliqz/react/modules/SearchEnginesModule.java

```diff
@@ -1,28 +1,96 @@
 package com.cliqz.react.modules;
 
-import com.facebook.react.bridge.Arguments;
+import android.util.Log;
+
 import com.facebook.react.bridge.Promise;
 import com.facebook.react.bridge.ReactApplicationContext;
 import com.facebook.react.bridge.ReactContextBaseJavaModule;
 import com.facebook.react.bridge.ReactMethod;
 import com.facebook.react.bridge.WritableArray;
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.bridge.WritableNativeArray;
+import com.facebook.react.bridge.WritableNativeMap;
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+
+import org.mozilla.gecko.EventDispatcher;
+import org.mozilla.gecko.util.BundleEventListener;
+import org.mozilla.gecko.util.EventCallback;
+import org.mozilla.gecko.util.GeckoBundle;
 
 /**
  * Copyright © Cliqz 2019
  */
-public class SearchEnginesModule extends ReactContextBaseJavaModule {
+public class SearchEnginesModule extends ReactContextBaseJavaModule implements BundleEventListener {
+    private Promise mCurrentPromise;
+    private ReactApplicationContext mReactContext;
+
     public SearchEnginesModule(ReactApplicationContext reactContext) {
         super(reactContext);
+        mReactContext = reactContext;
     }
 
     @Override
     public String getName() {
-        return "SearchEngines";
+        return "SearchEnginesModule";
+    }
+
+    @Override
+    public void initialize() {
+        EventDispatcher.getInstance().registerUiThreadListener(this, "SearchEngines:Data");
+    }
+
+    @Override
+    public boolean canOverrideExistingModule() {
+        return false;
+    }
+
+    @Override
+    public void onCatalystInstanceDestroy() {
+        EventDispatcher.getInstance().unregisterUiThreadListener(this, "SearchEngines:Data");
+    }
+
+    @Override
+    public void handleMessage(String event, GeckoBundle data, EventCallback callback) {
+        switch (event) {
+            case "SearchEngines:Data":
+                if (mCurrentPromise == null) {
+                    return;
+                }
+
+                final GeckoBundle[] engines = data.getBundleArray("searchEngines");
+                final WritableArray enginesArray = new WritableNativeArray();
+                // the first engine in the list is considered default
+                Boolean isDefaultSetOnce = true;
```

Comment:
In Java, using primitive `boolean` is preferred over the wrapper object `Boolean`.

https://stackoverflow.com/questions/18953251/when-should-i-use-boolean-instead-of-boolean
https://softwareengineering.stackexchange.com/questions/203970/when-to-use-primitive-vs-class-in-java